internal/mcpproxy/authorization.go

```diff
@@ -0,0 +1,239 @@
+	program     cel.Program
+}
+
+// compileAuthorization compiles the MCPRouteAuthorization into a compiledAuthorization for efficient CEL evaluation.
+func compileAuthorization(auth *filterapi.MCPRouteAuthorization) (*compiledAuthorization, error) {
+	if auth == nil {
+		return nil, nil
+	}
+
+	env, err := cel.NewEnv(
+		cel.Variable("args", cel.DynType),
+		cel.OptionalTypes(),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create CEL environment: %w", err)
+	}
+
+	compiled := &compiledAuthorization{
+		ResourceMetadataURL: auth.ResourceMetadataURL,
+	}
+
+	for _, rule := range auth.Rules {
+		cr := compiledAuthorizationRule{
+			Source: rule.Source,
+		}
+		for _, tool := range rule.Target.Tools {
+			ct := compiledToolCall{
+				BackendName: tool.BackendName,
+				ToolName:    tool.ToolName,
+			}
+			if tool.Arguments != nil && strings.TrimSpace(*tool.Arguments) != "" {
+				expr := strings.TrimSpace(*tool.Arguments)
+				ast, issues := env.Compile(expr)
+				if issues != nil && issues.Err() != nil {
+					return nil, fmt.Errorf("failed to compile arguments CEL for tool %s/%s: %w", tool.BackendName, tool.ToolName, issues.Err())
+				}
+				program, err := env.Program(ast, cel.CostLimit(10000), cel.EvalOptions(cel.OptOptimize))
+				if err != nil {
+					return nil, fmt.Errorf("failed to build arguments CEL program for tool %s/%s: %w", tool.BackendName, tool.ToolName, err)
+				}
+				ct.Expression = expr
+				ct.program = program
+			}
+			cr.Target = append(cr.Target, ct)
+		}
+		compiled.Rules = append(compiled.Rules, cr)
+	}
+
+	return compiled, nil
+}
+
+// authorizeRequest authorizes the request based on the given MCPRouteAuthorization configuration.
+
+func (m *MCPProxy) authorizeRequest(authorization *compiledAuthorization, headers http.Header, backendName, toolName string, arguments any) (bool, []string) {
+	if authorization == nil {
+		return true, nil
+	}
+
+	// If no rules are defined, deny all requests.
+	if len(authorization.Rules) == 0 {
+		return false, nil
+	}
+
+	// If the rules are defined, a valid bearer token is required.
+	token, err := bearerToken(headers.Get("Authorization"))
+	// This is just a sanity check. The actual JWT verification is performed by Envoy before reaching here, and the token
+	// should always be present and valid.
+	if err != nil {
+		m.l.Info("missing or invalid bearer token", slog.String("error", err.Error()))
+		return false, nil
+	}
+
+	claims := jwt.MapClaims{}
+	// JWT verification is performed by Envoy before reaching here. So we only need to parse the token without verification.
+	if _, _, err := jwt.NewParser().ParseUnverified(token, claims); err != nil {
+		m.l.Info("failed to parse JWT token", slog.String("error", err.Error()))
+		return false, nil
+	}
+
+	scopeSet := sets.New(extractScopes(claims)...)
+	var requiredScopesForChallenge []string
+
+	for _, rule := range authorization.Rules {
+		if !m.toolMatches(backendName, toolName, rule.Target, arguments) {
+			continue
```

Comment:
If I understand correctly, with the current logic, if the MCPRoute has, say 20 tools, and I just want to limit access so that for one particular tool I can not use a particular argument (e.g. you can't call the filesystem tool passing "/root" as an argument), then I could have to create 20 auth rules to allow all tools, and then the rule with the argument filter for that particular rule, right?

Is this the UX we want?
Should we consider bringing back the default action? Having a default allow and explicit denies can also help when MCP servers add more tools, etc (and we can consider that with the existing "tool filtering" we're only exposing the tools that we want to be called... so having to always list all allowed tools if you have any rule could not be the best UX.

WDYT?

Reply:
That's a good point. I think we should bring the default action back. Supporting wildcard is not enough here - the tool names may not share a similar name pattern.

Reply:
This can be achieved with a `Deny` rule for restricted folders and a `Allow` rule with the required sopes.
```yaml
authorization:
  defaultAction: Deny
  rules:
    - action: Deny
      target:
        tools:
          - backend: backend1
            toolName: listFiles
            arguments: 'args.folder == "restricted"'

    - action: Allow
      source:
        jwt:
          scopes:
            - read
      target:
        tools:
          - backend: backend1
            toolName: listFiles
```